feat: Client provider guidelines (#14)
open-feature/protocol: guideline/static-context-provider.md

  - In the constructor, the provider should check if the `baseURL` is a valid URL and return an error if the URL is invalid.
- `headers`: The headers to use when calling the OFREP endpoints *(e.g.:`Authorization`, Custom headers, etc ...)*.
- `pollInterval`: The polling interval defining how often to update the cached flags in the provider.
  - If `pollInterval` is equal to 0, polling will be disabled.
  - See [polling section](#polling) for more details.

## Initialize the provider
The following describes the [initialization](https://openfeature.dev/specification/sections/providers#24-initialization) of an OFREP provider:

The `initialize()` function must follow those steps:
1. Make a POST request to the `/ofrep/v1/evaluate/flags` endpoint with the evaluation context in the body.

   **Request body example**:
   ```json
    {
      "context": {
        "targetingKey": "f021d0f9-33b7-4b22-b0bd-9fec66ba1d7d",
        "firstname": "foo",
        "lastname": "bar",
        "email": "foo.bar@ofrep.dev"
      }
    }
   ```

    - If the endpoint returns an error, the `initialize()` function must error and exit.  
    - If the request is successful:

>       - Thew provider must store in a local cache all of the flags evaluation details returned by the API. 
>       - Thew provider must store the `ETag` header in the provider to be able to send it back later.

Reviewer: ```suggestion
      - The provider must store in a local cache all of the flags evaluation details returned by the API. 
      - The provider must store the `ETag` header in the provider to be able to send it back later.
```

> 2. If `pollInterval` is more than `0`, we should start a polling mechanism to periodically check for flag changes *([See polling section](#polling))*.

Reviewer 2: Should we define if I send an invalid `pollInterval`?

Author: I am not sure what you mean by an invalid `pollInterval`?

Reviewer 2: When someone sends a `pollInterval` of negative or zero.

> 2. If `pollInterval` is more than `0`, we should start a polling mechanism to periodically check for flag changes *([See polling section](#polling))*.

Author: I've adapted the language style, I found it better now.

When calling an evaluation function, the provider must follow those steps:
1. Retrieve the evaluation details for this flag from the local cache.
2. If the provider is not able to find the flag in the local cache, it must return an `FLAG_NOT_FOUND` error.
3. If the evaluation details contains an error, the provider must map the error in provider to an OpenFeature error and return it.
4. If the value of the evaluation details in the cache has a different type than the one expected by the evaluation function the provider must return a `TYPE_MISMATCH` error.
5. If the evaluation details retrieved from the cache is successful, the provider returns the evaluation details.

```mermaid
flowchart TD
    A[evaluation function] --> D{Does the provider have an entry in cache for this flag key ?}
    D --> |NO| E(return a FLAG_NOT_FOUND error)
    D --> |YES| F{Is cached evaluation details in error?} 
    F --> |YES| G(Map the error as an OpenFeature provider error and return it)
    F --> |NO| H{Are the flag and evaluation return types the same?}
    H --> |YES| I(return the evaluation response)
    H --> |NO| J(return a TYPE_MISMATCH error)
```

## Polling

> The polling system will make a `POST` request periodically to the `/ofrep/v1/evaluate/flags` endpoint with the evaluation context in the body to check if there is a change in the flags evaluation to be able to store it.
> The goal is to be able to know when a flag has changed for this evaluation context.

Reviewer: ```suggestion
The polling system must periodically (interval defined by `pollingInterval`) make a `POST` request to the `/ofrep/v1/evaluate/flags` endpoint with the evaluation context in the body to check if there is a change in the flags evaluation to be able to store it.
If the flags changed, the provider cache for evaluation is updated.
```

When calling the API the provider can receive those response codes:
- `304`: Means that the current cache is up-to-date.
- `401`, `403`: The provider is not authorized to call the OFREP API. In that situation the provider should emit a `fatal` event and stop the polling mechanism. The provider will be in a [terminal error state](https://openfeature.dev/specification/sections/flag-evaluation#requirement-177).
- `429`: The provider has reached the rate limit of the flag management system. In that situation the provider should read the `Retry-After` header from the response and ensure that no call to the endpoint happened before this date.
- `200`: The cache is outdated. In that situation the provider must:
  1. Clear the actual cache.
  2. Cache the new evaluation details received.
  3. Read the `ETag` header and store it for future call the API.
  4. Emit an `ConfigurationChanged` event containing in the `flagsChanged` field, a list of the flag key that have changed.

## Change context

> In the client providers, a change of context should be handle to retrieve a new version of the cache, it is called that the [Provider context reconciliation](https://openfeature.dev/specification/sections/providers/#26-provider-context-reconciliation).

Reviewer: I do not fully get this sentence.

**Pre-requisite:**
- Understanding of [general provider concepts](https://openfeature.dev/docs/reference/concepts/provider/)
- Understanding of the [OFREP](../../README.md)
- Understanding of the [OFREP OpenAPI specification](../../service/openapi.yaml)

## Configuration
An OFREP client provider implementation must, at the time of creation, accept at least these options
- `baseURL`: The base URL of the [flag management system](https://openfeature.dev/specification/glossary#flag-management-system).  
  This must be the base of the URL pointing before the `/ofrep` namespace of the API.
  - In the constructor, the provider should check if the `baseURL` is a valid URL and return an error if the URL is invalid.
- `headers`: The headers to use when calling the OFREP endpoints *(e.g.:`Authorization`, Custom headers, etc ...)*.
- `pollInterval`: The polling interval defining how often to update the cached flags in the provider.
  - If `pollInterval` is equal to 0, polling will be disabled.
  - See [polling section](#polling) for more details.

> - `timeout`: This specifies the duration to wait for a response before canceling the HTTP request. If no value is provided, a default timeout of `10 seconds` must be applied.

Reviewer 3: ```suggestion
- `timeoutMs`: This specifies the duration to wait for a response before canceling the HTTP request. If no value is provided, a default timeout of `10_000 milliseconds` must be applied.
```

It's called `timeoutMs` already in the JavaScript providers.

We will keep the document language agnostic.

**Pre-requisite:**
- Understanding of [general provider concepts](https://openfeature.dev/docs/reference/concepts/provider/)
- Understanding of the [OFREP](../../README.md)
- Understanding of the [OFREP OpenAPI specification](../../service/openapi.yaml)

## Constructor
An implementation of an OFREP client provider should allow in the creation of the provider to take as options:
- `baseURL`: The base URL of the [flag management system](https://openfeature.dev/specification/glossary#flag-management-system). This should be the base of the URL pointing before the `/ofrep` namespace of the API.
- `headers`: The headers to use when calling the OFREP endpoints *(ex:`Authorization`, Custom headers, etc ...)*.
- `pollInteral`: The polling interval defining when to call again the flag management system.

In the constructor, the provider should check if the `baseURL` is a valid URL and return an error if the URL is invalid.

> ## Initialize the provider

Reviewer 4: I'm a bit torn here wether my thinking is going to deep or the use case is too fringe but bare with me:

Some applications have strict rules or goals of how fast they need to be able to show their initial screen. These apps may not have time (or do not want) to wait for network requests to finish before considering a feature flagging SDK to be initialised.

Obviously there are many ways one can solve this, but we have written mobile providers with disk persistence where we then can pass a `FetchingStrategy` to the constructor where the app owner can decide wether or not they want the `initialize()` function to block until fresh data is available or simply load any previous data from disk (and then refresh from the backend in the background). 

Firebase Remote Config explains the concepts [here](https://firebase.google.com/docs/remote-config/loading).

Reviewer 3: I like the idea of various initialization strategies and the link to Firebase Remote Config is a great resource.

Perhaps one way to implement that would be to put this section under a "Fetch on load" strategy. We could start with a single strategy (and perhaps make it the default) but that would give us the ability to add additional strategies to support specific use cases.

Author: Interesting point here. I see the value of having multiple fetching strategies but I am not sure how to frame that into this document.

I see 2 options:
1. We try to define the different possible ways of doing the initialisation for the OFREP client providers, and we add them to the document.
2. We remove that block since there is some chance that it could be different to initialize something depending on the platform (web/mobile).

Do you have any preference?

Reviewer 4: I'm thinking that it might be most reasonable to ignore my comment about this and go ahead with a blocking network call as part of the initialize function in order to not block this document.

If we believe that the topic warrants additional discussion we can open an issue to discuss it there and amend the spec or add an appendix. wdyt?

Author: As decided in the last OFREP meeting, we will get rid of the `/configuration` endpoint.
So I will resolve this conversation, because we won't have issue on how fast we can start.

An implementation of an OFREP client provider should start with an initialization of the provider.

The `initialize()` function should follow those steps:

> 1. Make a GET request to the `/ofrep/v1/configuration` endpoint to retrieve the configurations return by the flag management system and store them in memory to be available for all the function of the provider. *(See [Annexe 1](#annexe-1) for the description of the endpoint response)*

Reviewer 3: Can this step be optional? The provider should be able to work without having to call the configuration API.

Reviewer: This is also my general feeling, that this should be an opt-in feature as @beeme1mr says.

Author: I was more thinking toward something that can be enabled by default and opt-out on demand.
See https://github.com/open-feature/protocol/pull/14#discussion_r1601965901

Reviewer 3: No longer applicable

2. Make a POST request to the `/ofrep/v1/evaluate/flags` endpoint with the evaluation context in the body.  
   - If the endpoint returns an error, the `initialize()` function must error and exit.  

>    - If the request is successful, we should store in a local cache all of the flags evaluation results returned by the API in a local cache. We should also store the `ETag` header in the provider to be able to send it back later.

Reviewer: I think it would be better to leave the `we` out and talk in third person about the provider.

> 3. If polling is enabled, the function should start the polling loop *(See [polling section](#polling))*.

Reviewer: ```suggestion
3. If polling is enabled, the polling loop should start now *(See [polling section](#polling))*.
```

## Evaluation
The evaluation should not perform any remote API calls.

When calling an evaluation function the provider should check if the associated type is supported, by checking the key `capabilities.flagEvaluation.unsupportedTypes` from the configuration endpoint.

> - If the type is unsupported, we should exit early and directly return an error.

Reviewer: Do we know which error we want to throw here? I think it would be nice to specify here.
Maybe we can even have a new Error type, the case that a system does not support a type happens pretty often. (independently from OFREP)

Author: Yes, this is true, should we add that to the OF spec?

Reviewer: Mh yes I think we should add it to the OF spec.
I could open a PR next days, what do you think @thomaspoignant ?

Author: It would be awesome

- If the type is supported, we should retrieve the flag from the local cache of the flags evaluation.
  - If you are not able to find the flag, you should return an FlagNotFound error.
  - If the remote evaluation for this flag has return an error, you should map the error in provider and return the associated error.
  - If the value retrieve from the cache has a different type than the one expected you should return a TypeMismatch error.
  - If the cached evaluation is in success you should return the evaluation response.

```mermaid
flowchart TD
    A[evaluation\nfunction] --> B{Is function type in\nunsupportedTypes?}
    B --> |YES| C(return an error)
    B --> |NO| D{Is flag key stored\nin local cache?}
    D --> |NO| E(return a FlagNotFound error)
    D --> |YES| F{Is cached evaluation\nresponse in error?} 
    F --> |YES| G(Map the error as a\nprovider error and return)
    F --> |NO| H{Is the flag value the\nsame type as the\nevaluation function?}
    H --> |YES| I(return the evaluation response)
    H --> |NO| J(return a TypeMismatch error)

> ```

Reviewer: Love it! :)

Reviewer 4: Yes, this is excellent!

## Polling

> The polling system will make a POST request periodically tp the `/ofrep/v1/evaluate/flags` endpoint to check if there is a change in the flags evaluation to be able to store it.

Reviewer 3: ```suggestion
The polling system will make a POST request periodically to the `/ofrep/v1/evaluate/flags` endpoint to check if there is a change in the flags evaluation to be able to store it.
```

> If an `ETag` is available we should always add the header `If-None-Match` with the `ETag` value.
> - If the cache is still up-to-date we will receive a `304` telling us that the nothing has changed on the flag management system side.
> - If the cache is outdated we will receive a `200` with the new values of all the flags. In that situation we should:
>   1. Replace the actual local cache of flags evaluations.
>   2. Store the new `ETag` value for the future call.

Reviewer: ```suggestion
If an `ETag` of a former evaluation is available we should always add the header `If-None-Match` with the `ETag` value.
- If the cache is still up-to-date we will receive a `304` telling us that the nothing has changed on the flag management system side.
- If the cache is outdated we will receive a `200` with the new values of all the flags. In that situation we should:
  1. Replace the actual local cache of flags evaluations.
  2. Store the new `ETag` value for the future call.
```

## Annexe 1
**`/ofrep/v1/configuration` description:**
The endpoint will return a list of configurations for the provider:
- `name`: Name of the flag management system, it should be used in the `metadata.name` name (ex: `OFREP Web Provider ${metadata.name}`).
- `capabilities`: List of capabilities of the flag management system and their associated configuration. *Check the [capabilities section](#capabilities) for the details of each capability.*

### Capabilities
In this section we will describe each capability and describe their default value and usage.

#### Cache Invalidation
The capability `cacheInvalidation` describes how the mechanism of cache invalidation works.

##### Polling
`polling`: define how the provider should do the polling

> - `enabled`: if `true` the provider should poll the `/ofrep/v1/evaluate/flags` regularly to check if any flag evaluation has changed in the flag management system.

Reviewer 3: This isn't the current behavior in the OFREP web provider.

https://github.com/open-feature/js-sdk-contrib/blob/main/libs/providers/ofrep-web/src/lib/ofrep-web-provider.ts#L90

Author: OFREP web provider by default is doing polling.
You have a default value set if no `pollingInterval` is provided during the initialization.

https://github.com/open-feature/js-sdk-contrib/blob/c00f998227f55ca5a64b01d039ab6ec2aff18f7c/libs/providers/ofrep-web/src/lib/ofrep-web-provider.ts#L78
